File: lc_soln/0239_sliding_window_maximum/solution.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        std::deque<int> q;
        std::vector<int> result;
        int start = 0;
        for (int i = 0; i < nums.size(); ++i) {
            while (!q.empty() && nums[q.back()] < nums[i]) {
                q.pop_back();
            }
            q.push_back(i);
            if (q.front() < start) {
                q.pop_front();
            }
            if (i - start + 1 >= k) {
                result.push_back(nums[q.front()]);
                start++;
            }
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0239_sliding_window_maximum/solution.rs
struct Solution;

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut q: VecDeque<usize> = VecDeque::new();
        let mut result: Vec<i32> = Vec::new();
        let mut start = 0;
        for (i, &num) in nums.iter().enumerate() {
            while let Some(&b) = q.back() && nums[b] < num {
                q.pop_back();
            }
            q.push_back(i);
            if let Some(&f) = q.front() && start > f {
                q.pop_front();
            }
            if (i - start + 1) as i32 >= k {
                result.push(nums[q[0]]);
                start += 1;
            }
        }
        result
    }
}

fn main() {
}
